NOnion/TorCircuit.fs

```diff
@@ -0,0 +1,219 @@
+﻿namespace NOnion
+
+open System.Security.Cryptography
+open System.Threading
+
+open NOnion.Cells
+open NOnion.Crypto
+open NOnion.Crypto.Kdf
+open NOnion.Utility
+
+type TorCircuit
+    private
+    (
+        circuitId: uint16,
+        guard: TorGuard,
+        kdfResult: KdfResult
+    ) as self =
+
+    let cryptoState = TorCryptoState.FromKdfResult kdfResult
+    // Prevents multiple write from breaking the cryptoState because of RC
+    let sendLock: obj = obj ()
+
+    let streamMessages = Event<uint16 * RelayData> ()
+
+    let window = TorWindow Constants.DefaultCircuitLevelWindowParams
+
+    let mutable streamsCount: int = 1
+    // Prevents two stream setup happening at once (to prevent race condition on writing to StreamIds list)
+    let streamSetupLock: obj = obj ()
+
+    do
+        guard.MessageReceived
+        |> EventUtils.FilterByKey circuitId
+        |> Event.add self.HandleIncomingMessage
+
+    [<CLIEvent>]
+    member __.StreamMessages = streamMessages.Publish
+
+    member __.Id = circuitId
+
+    member __.Send (streamId: uint16) (relayData: RelayData) =
+        async {
+            Monitor.Enter sendLock
```

Comment:
why is this nnot using a lock statement instead of Monitor.Enter?

Reply:
because that's an async block and using lock statement on an async function is problamatic